### PR TITLE
Actionable Notifications

### PR DESCRIPTION
**Pull request by JBassett**
Current work in progress.  Basic functionality working.

Screenshot_20200201-234943

Example service call (supports up to 3 actions):
```
- service: notify.mobile_app_pixel_2
  data:
    message: 'Something really cool happened'
    title: 'Action Time'
    data:
      data:
        action_1_key: 'EVENT_KEY'
        action_1_title: 'Button Text'
```

Example event in HA:
```
{
    "event_type": "mobile_app_notification_action",
    "data": {
        "action": "KEY_ONE",
        "action_2_title": "Action 2",
        "action_3_title": "Action 3",
        "action_1_title": "Action 1",
        "action_1_key": "KEY_ONE",
        "action_2_key": "KEY_TWO",
        "action_3_key": "KEY_THREE"
    },
    "origin": "REMOTE",
    "time_fired": "2020-02-02T04:45:05.550251+00:00",
    "context": {
        "id": "abc123",
        "parent_id": null,
        "user_id": "123abc"
    }
}
```

Example Consumption in Automation:
```
- id: '1523975760898'
  alias: '[Notification - Action] Close Garage'
  initial_state: 'on'
  trigger:
    - platform: event
      event_type: mobile_app_notification_action
      event_data:
        action: KEY_ONE
  condition: []
  action:
    - service: cover.close_cover
      data:
        entity_id: cover.garage_door_opener

```

### CODE REVIEW

**Comment by SamJongenelen** [comment]
Solves #176

**Comment by dshokouhi** [comment]
This is great! Would it be possible to update the description to include the automation trigger example too?  I tried looking on the iOS side but seems for android it will be different.  Thank you for including examples for the rest of it :)

https://companion.home-assistant.io/en/notifications/actionable#building-automations-for-notification-actions

**Comment by JBassett** [comment]
@dshokouhi done.  If someone wanted to update the actual documentation for android I wouldn't be upset... ;-)
I looked into matching iOS for the actions but they were not compatible.

File: app/src/main/java/io/homeassistant/companion/android/notifications/NotificationActionReceiver.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    override fun onReceive(context: Context?, intent: Intent?) {
+        DaggerServiceComponent.builder()
+            .appComponent((context!!.applicationContext as GraphComponentAccessor).appComponent)
+            .build()
+            .inject(this)
+
+        intent?.getParcelableExtra<NotificationAction>(EXTRA_NOTIFICATION_ACTION)?.let {
+            ioScope.launch {
+                integrationUseCase.fireEvent(it.key, it.data)
```
Comment (balloob) [inline app/src/main/java/io/homeassistant/companion/android/notifications/NotificationActionReceiver.kt]: We should not allow the user to pick the event name to fire. Instead, we should pick an event name like "mobile_app_action" and pass both the key and data as event data.

Comment (balloob) [reply app/src/main/java/io/homeassistant/companion/android/notifications/NotificationActionReceiver.kt]: Actually, why do we include `it.data` ? We should drop that.

Comment (JBassett) [reply app/src/main/java/io/homeassistant/companion/android/notifications/NotificationActionReceiver.kt]: We are basically ecoing back the data that the user passed in the notification.  You can put any string->string mapping in there.  We could limit it to a standard event name and only pass the key if you want.  I was just trying to leave it more open, I think the iOS app echos all data back too.

Comment (balloob) [reply app/src/main/java/io/homeassistant/companion/android/notifications/NotificationActionReceiver.kt]: ok

Comment (JBassett) [reply app/src/main/java/io/homeassistant/companion/android/notifications/NotificationActionReceiver.kt]: Which option would you like?

Comment (JBassett) [reply app/src/main/java/io/homeassistant/companion/android/notifications/NotificationActionReceiver.kt]: I updated to use constant event type and added an "action" element to the event data that is the selected option.

Comment (balloob) [reply app/src/main/java/io/homeassistant/companion/android/notifications/NotificationActionReceiver.kt]: I'm fine with echo'ing it back.

**Comment by SamJongenelen** [comment]
Thanks guys, awesome

**Comment by bransjr** [comment]
Hey, I'm using your example yaml verbatim and get a notification that it is "Action time" and that "something really cool happened" but there are not buttons. any ideas what I'm doing wrong? Might it be a permission with the app?

**Comment by Chumba** [comment]
@bransjr I'm having the same result, did you ever get it sorted out?

**Comment by dshokouhi** [comment]
@bransjr  and @Chumba lets not discuss in a merged PR but please make sure to check the companion docs for the updated example: https://companion.home-assistant.io/docs/notifications/actionable-notifications#building-automations-for-notification-actions

Scroll down for the android example

**Pull request merged**